File: libs/api/src/main/java/com/example/api/HttpClientWithPinningProvider.java
package com.example.api;

import java.security.GeneralSecurityException;
import java.security.KeyStore;
import java.util.Collections;

import javax.net.ssl.TrustManager;
import javax.net.ssl.TrustManagerFactory;
import javax.net.ssl.X509TrustManager;

import okhttp3.ConnectionSpec;
import okhttp3.OkHttpClient;
import okhttp3.TlsVersion;

public class HttpClientWithPinningProvider {
  private static final OkHttpClient CLIENT = init();

  private static OkHttpClient init() {
    ConnectionSpec.Builder spec = new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
      .tlsVersions(TlsVersion.TLS_1_2);
    OkHttpClient.Builder client = new OkHttpClient.Builder()
      .connectionSpecs(Collections.singletonList(spec.build()));
    // This setup is necessary because API 16-18 support TLS 1.2,
    // but use 1.0 as default.
    setupTls(client);
    setupPinning(client);

    return client.build();
  }

  private static void setupPinning(OkHttpClient.Builder client) {
    // Implement certificate pinning here
  }

  private static void setupTls(OkHttpClient.Builder builder) {
    try {
      TrustManagerFactory factory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
      factory.init((KeyStore) null);
      for (TrustManager trustManager : factory.getTrustManagers()) {
        if (trustManager instanceof X509TrustManager) {
          builder.sslSocketFactory(new Tls12SslSocketFactory(), (X509TrustManager) trustManager);
          break;
        }
      }
    } catch (GeneralSecurityException e) {
      throw new RuntimeException(e);
    }
  }

  public static OkHttpClient getClient() {
    return CLIENT;
  }
}

File: libs/api/src/main/kotlin/com/example/api/HttpClientWithPinningProvider.kt
package com.example.api

import okhttp3.ConnectionSpec
import okhttp3.OkHttpClient
import okhttp3.TlsVersion
import java.security.GeneralSecurityException
import java.security.KeyStore
import javax.net.ssl.TrustManagerFactory
import javax.net.ssl.X509TrustManager

object HttpClientWithPinningProvider {
  val client = init()

  private fun init(): OkHttpClient {
    val spec = ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
      .tlsVersions(TlsVersion.TLS_1_2)
    val client = OkHttpClient.Builder()
      .connectionSpecs(listOf<ConnectionSpec>(spec.build()))
    // This setup is necessary because API 16-18 support TLS 1.2,
    // but use 1.0 as default.
    setupTls(client)
    setupPinning(client)

    return client.build()
  }

  private fun setupPinning(client: OkHttpClient.Builder) {
    // Implement certificate pinning here
  }

  private fun setupTls(builder: OkHttpClient.Builder) {
    try {
      val factory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
      factory.init(null as KeyStore?)
      for (trustManager in factory.trustManagers) {
        if (trustManager is X509TrustManager) {
          builder.sslSocketFactory(Tls12SslSocketFactory(), trustManager)
          break
        }
      }
    } catch (e: GeneralSecurityException) {
      throw RuntimeException(e)
    }
  }
}
